spec/abci++/abci++_app_requirements_002_draft.md

```diff
@@ -177,3 +179,818 @@ Likewise, `ExtendVote` can also be non-deterministic:
+deterministic since, given a transaction, nodes' Applications
+might have a different `CheckTxState` when they receive it and check their validity via `CheckTx`.
+It is
+semantically meaningless to Tendermint.
+
+`Events` include any events for the execution, though since the transaction has not
+been committed yet, they are effectively ignored by Tendermint.
+
+In v0.35.x and later, there is a `Priority` field in `ResponseCheckTx` that can be
+used to explicitly prioritize transactions in the mempool for inclusion in a block
+proposal.
+
+#### Specifics of `ExecTxResult`
+
+`FinalizeBlock` is the workhorse of the blockchain. Tendermint delivers the decided block,
+including the list of all its transactions synchronously to the Application.
+The block delivered (and thus the transaction order) is the same at all correct nodes as guaranteed
+by the Agreement property of Tendermint consensus.
+
+In same block execution mode `ExecTxResult` is returned by `PrepareProposal` so that its data can
+be used to build the block to be proposed in the current height.
+
+The `Data` field in `ExecTxResult` contains an array of bytes with the transaction result.
+It must be strictly deterministic (all nodes should return the same value), but can contain arbitrary
+data. Likewise, the value of `Code` must be deterministic.
+Both the `Code` and `Data` are included in a structure that is hashed into the
+`LastResultsHash` of the block header in the next height (next block execution mode), or the
+header of the block to propose in the current height (same block execution mode, `ExecTxResult` as
+part of `PrepareProposal`).
+
+If `ExecTxResult` contains `Code != 0`, the transaction will be considered invalid,
+though it is still included in the block.
+
+`Events` include any events for the execution, which Tendermint will use to index
+the transaction by. This allows transactions to be queried according to what
+events took place during their execution.
+
+### Updating the Validator Set
+
+The application may set the validator set during
+[`InitChain`](./abci%2B%2B_methods_002_draft.md#initchain), and may update it during
+[`FinalizeBlock`](./abci%2B%2B_methods_002_draft.md#finalizeblock)
+(next block execution mode) or
+[`PrepareProposal`](./abci%2B%2B_methods_002_draft.md#prepareproposal)/[`ProcessProposal`](./abci%2B%2B_methods_002_draft.md#processproposal)
+(same block execution mode). In all cases, a structure of type
+[`ValidatorUpdate`](./abci%2B%2B_methods_002_draft.md#validatorupdate) is returned.
+
+The `InitChain` method can return a list of validators.
+If the list is empty, Tendermint will use the validators loaded in the genesis
+file.
+If the list returned by `InitChain` is not empty, Tendermint will use its contents as the validator set.
+This way the application can set the initial validator set for the
+blockchain.
+
+Note that the maximum total power of the validator set is bounded by
```

Comment:
Suggestion: use a different format here, start with a "Note" or "Important".

Reply:
Suggestion: re-organize the description so that users know that voting power are "arbitrary", but that the maximum power for a set is bounded. I think they are unlikely to violate this requirement though.

Reply:
OK, let me see what I can do